crates/ruma-signatures/src/functions.rs

```diff
@@ -572,39 +575,44 @@ pub fn verify_event(
             CanonicalJsonValue::Object(hashes) => match hashes.get("sha256") {
                 Some(hash_value) => match hash_value {
                     CanonicalJsonValue::String(hash) => hash,
-                    _ => return Err(Error::new("sha256 hash must be a JSON string")),
+                    _ => return Err(JsonError::not_of_type("sha256 hash", JsonType::String)),
                 },
-                None => return Err(Error::new("field `hashes` must be a JSON object")),
+                None => return Err(JsonError::not_of_type("hashes", JsonType::Object)),
             },
-            _ => return Err(Error::new("event missing sha256 hash")),
+            _ => return Err(JsonError::field_missing_from_object("sha256")),
         },
-        None => return Err(Error::new("field `hashes` must be present")),
+        None => return Err(JsonError::field_missing_from_object("hashes")),
     };
 
     let signature_map = match object.get("signatures") {
         Some(CanonicalJsonValue::Object(signatures)) => signatures,
-        Some(_) => return Err(Error::new("field `signatures` must be a JSON object")),
-        None => return Err(Error::new("JSON object must contain a `signatures` field.")),
+        Some(_) => return Err(JsonError::not_of_type("signatures", JsonType::Object)),
+        None => return Err(JsonError::field_missing_from_object("signatures")),
     };
 
     let servers_to_check = servers_to_check_signatures(object, version)?;
-    let canonical_json = from_json_str(&canonical_json(&redacted))?;
+    let canonical_json =
+        from_json_str(&canonical_json(&redacted)).map_err(|e| JsonError::from(e))?;
 
     for entity_id in servers_to_check {
         let signature_set = match signature_map.get(entity_id.as_str()) {
             Some(CanonicalJsonValue::Object(set)) => set,
-            Some(_) => return Err(Error::new("signatures sets must be JSON objects")),
-            None => {
-                return Err(Error::new(format!("no signatures found for entity `{}`", entity_id)))
+            Some(_) => {
+                return Err(JsonError::not_multiples_of_type("signature sets", JsonType::Object))
             }
+            None => return Err(VerificationError::signature_not_found(entity_id)),
         };
 
-        let mut maybe_signature = None;
-        let mut maybe_public_key = None;
+        struct SignatureAndPubkey<'a> {
+            signature: &'a CanonicalJsonValue,
+            public_key: &'a String,
+        }
+
+        let mut maybe = None;
```

Comment:
Previously this section had 2 `PublicKeyNotFound` errors in two different spots, I saw the existence of them both being `Some` or `None` were linked, and made it a bit neater.

Reply:
Maybe this could be renamed to `maybe_signature_and_public_key` (or something similar)?